File: src/harness/spawn-server.ts
import { spawn } from "node:child_process";
import { EngineStartTimeoutError } from "../errors.js";

export interface SpawnServerOptions {
  /** Engine name, used only for error messages. */
  engine: string;
  /** Executable name or path -- never a shell string. */
  command: string;
  /** Argv array -- never interpolated into a shell string (command injection safety). */
  args: string[];
  /** URL to poll until it responds successfully, indicating the server is ready. */
  readyCheckUrl: string;
  /** Max time to wait for readyCheckUrl to respond before giving up. */
  timeoutMs: number;
  /** Poll interval while waiting for the server to become ready. */
  pollIntervalMs?: number;
  verbose?: boolean;
}

export interface SpawnedServer {
  process: ReturnType<typeof spawn>;
  stop(): void;
}

/**
 * Spawns a long-running server process (never through a shell -- argv array
 * only, to avoid command-injection risk) and polls a URL until it responds,
 * so callers get a server that is actually ready to accept requests rather
 * than racing against startup.
 */
export async function spawnServerAndWaitReady(
  opts: SpawnServerOptions,
): Promise<SpawnedServer> {
  const pollIntervalMs = opts.pollIntervalMs ?? 500;
  const child = spawn(opts.command, opts.args, {
    stdio: opts.verbose ? "inherit" : "ignore",
  });

  const stop = () => {
    if (!child.killed) {
      child.kill();
    }
  };

  const deadline = Date.now() + opts.timeoutMs;
  let lastError: unknown;
  while (Date.now() < deadline) {
    if (child.exitCode !== null) {
      throw new Error(
        `${opts.engine}: process exited early (code ${child.exitCode}) before becoming ready`,
      );
    }
    try {
      const res = await fetch(opts.readyCheckUrl, {
        signal: AbortSignal.timeout(2000),
      });
      if (res.ok) {
        return { process: child, stop };
      }
    } catch (err) {
      lastError = err;
    }
    await new Promise((resolve) => setTimeout(resolve, pollIntervalMs));
  }

  stop();
  const timeoutErr = new EngineStartTimeoutError(opts.engine, opts.timeoutMs);
  if (lastError instanceof Error) {
    timeoutErr.cause = lastError;
  }
  throw timeoutErr;
}

File: python/src/inferbench/harness/spawn_server.py
"""
Ported from src/harness/spawn-server.ts. Spawns a long-running engine
server process and polls a URL until it responds, so callers get a server
that is actually ready rather than racing against startup.

Security note: `command` and `args` are always a fixed argv list passed to
subprocess.Popen with no shell involved (no shell=True anywhere in this
module or its callers) -- the model spec a user passes on --model becomes
a single argv element, never string-concatenated into a shell command, so
there is no command-injection path through it. This mirrors the same
argv-array discipline the original TypeScript harness documents in its own
comments.
"""
from __future__ import annotations

import subprocess
import time
import urllib.error
import urllib.request
from dataclasses import dataclass
from typing import Callable, List, Optional

from ..errors import EngineStartTimeoutError


@dataclass
class SpawnedServer:
    process: subprocess.Popen
    stop: Callable[[], None]


def spawn_server_and_wait_ready(
    *,
    engine: str,
    command: str,
    args: List[str],
    ready_check_url: str,
    timeout_ms: int,
    poll_interval_ms: int = 500,
    verbose: bool = False,
) -> SpawnedServer:
    stdio = None if verbose else subprocess.DEVNULL
    process = subprocess.Popen([command, *args], stdout=stdio, stderr=stdio)  # noqa: S603 -- fixed argv, no shell

    def stop() -> None:
        if process.poll() is None:
            process.kill()

    deadline = time.monotonic() + timeout_ms / 1000
    last_error: Optional[BaseException] = None

    while time.monotonic() < deadline:
        if process.poll() is not None:
            raise RuntimeError(
                f"{engine}: process exited early (code {process.returncode}) before becoming ready"
            )
        try:
            with urllib.request.urlopen(ready_check_url, timeout=2) as response:  # noqa: S310
                if 200 <= response.status < 300:
                    return SpawnedServer(process=process, stop=stop)
        except Exception as err:  # noqa: BLE001 -- polling loop; any failure just means "not ready yet"
            last_error = err
        time.sleep(poll_interval_ms / 1000)

    stop()
    timeout_err = EngineStartTimeoutError(engine, timeout_ms)
    if last_error is not None:
        timeout_err.__cause__ = last_error
    raise timeout_err
